server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -2919,45 +2920,65 @@ private void doSnapshotShard(SnapshotShardContext context) {
                 };
             }
 
+            // filesToSnapshot will be emptied while snapshotting the file. We make a copy here for cleanup purpose in case of failure.
+            final List<FileInfo> copyOfFilesToSnapshot = List.copyOf(filesToSnapshot);
```

Comment:
I wonder whether it'd be nicer to make this `AtomicReference<List<FileInfo>> filesToCleanUp` which we set to `null` when we pass the point of no return. The mix of try/catch blocks and `ActionListener.wrap` calls is kinda hard to follow IMO, I'd rather we did everything with `ActionListener` utils if possible.

Reply:
Great suggestion. It made the diff much nicer to look at. Thanks!